[inductor] Make codecache file permissions less restrictive

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100870

`tempfile.mkstemp` always creates the file `0o600` permissions, so
only the current user can access it. Instead, this salts the original
filename with the pid and thread id to avoid conflicts between
temporary files.

cc @soumith @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

peterbell10 committed:
[inductor] Make codecache file permissions less restrictive

`tempfile.mkstemp` creates a file with no permissions given to the
user's group or to other users on the machine. Whereas the default is
usually read-write permissions for group members and read-only for
other users.

This creates an issue in GitHub codespaces for example, where the
terminal runs under a different user from the vs code.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make codecache file permissions less restrictive"


`tempfile.mkstemp` creates a file with no permissions given to the
user's group or to other users on the machine. Whereas the default is
usually read-write permissions for group members and read-only for
other users.

This creates an issue in GitHub codespaces for example, where the
terminal runs under a different user from vs code.

cc soumith voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make codecache file permissions less restrictive"


`tempfile.mkstemp` always creates the file `0o600` permissions, so
only the current user can access it. Instead, this uses `open` in a
temporary sub-directory which lets the OS decide the file's permissions.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Make codecache file permissions less restrictive"


`tempfile.mkstemp` always creates the file `0o600` permissions, so
only the current user can access it. Instead, this salts the original
filename with the pid and thread id to avoid conflicts between
temporary files.

[ghstack-poisoned]

jansel commented on torch/_inductor/codecache.py:
```diff
@@ -230,15 +231,15 @@ def write(source_code, ext, extra=""):
-        tmp_path = pathlib.Path(tempdir) / path.name
-        with tmp_path.open("w") as f:
-            f.write(source_code)
+    tmp_path = path.parent / f".{path.name}.{salt}.tmp"
```
Don't include `path.name` here, we are getting pretty close to the filename length limit on some older file system types.

peterbell10 committed:
Update on "[inductor] Make codecache file permissions less restrictive"


`tempfile.mkstemp` always creates the file `0o600` permissions, so
only the current user can access it. Instead, this salts the original
filename with the pid and thread id to avoid conflicts between
temporary files.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge
